### ISSUE TITLE
.distinct() operator

### ISSUE THREAD

**Issue by rudijs** [ReactiveX/RxJS#1156]
Hi,

I'm new to RxJS and in one of the early chapters of my book the .distinct() operator is used in the examples.

I see it's not part of this library, I'm assuming that's for performance reasons.

What would be the recommended approach for emiting unique values only with RxJS?

Thank you.

**Comment by staltz** [comment]
@rudijs you can always keep a buffer with previous values, in a `scan` operator. I suppose `distinct` was removed because it is misleading, usually people want distinctUntilChanged, and distinct has memory from the whole history. If people really want it:

```js
source.scan((acc, curr) => {
  if (acc.indexOf(curr) === -1) { 
    acc.push(curr);
    return curr;
  } else {
    return 'ignore this';
  }
}, []).filter(x => x !== 'ignore this');
```

**Comment by bouzuya** [comment]
does it work?

**Comment by saneyuki** [comment]
@staltz I feel your example would work only once because the operator callback of `scan()` does not return `acc`

**Comment by staltz** [comment]
Right, I was mistaken, should have been:

```js
source.scan((acc, curr) => {
  if (acc.indexOf(curr) === -1) {
    acc.push(curr);
    acc.last = curr;
    return acc;
  } else {
    acc.last = 'ignore this';
    return acc;
  }
}, [])
.map(acc => acc.last)
.filter(last => last !== 'ignore this');
```
```

**Comment by trxcllnt** [comment]
@staltz @rudijs we haven't implemented `distinct` yet, but I believe it should be added to the list of operators we still need to implement. Even though it can be implemented by composing other operators, so can pretty much everything else, but it's a common enough operator that it should exist on its own cc: @blesh

**Comment by mattpodwysocki** [comment]
@trxcllnt @staltz I agree, that `distinct` should be an operator unto itself.  Also, due to its constraint size of the buffer, we might want to add a flushing feature else the inner hash is going to grow infinitely.

**Comment by justinwoo** [comment]
I'd be up for writing this operator up. Would it make sense for there to be an optional "flushing observable" that gets sent in for triggering hash flushing for every event?

**Comment by blesh** [comment]
Yes, this should be implemented, but probably included only in KitchenSink, since it can be accomplished easily with other common operators.

**Comment by kwonoj** [comment]
This operator's landed via https://github.com/ReactiveX/RxJS/pull/1180, closing issue.
